spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala

```diff
@@ -324,46 +355,60 @@ case class DeltaParquetFileFormat(
           // Create vectors for all needed metadata columns.
           // We can't use the one from Parquet reader as it set the
           // [[WritableColumnVector.isAllNulls]] to true and it can't be reset with using any
           // public APIs.
           trySafely(useOffHeapBuffers, size, metadataColumns) { writableVectors =>
             val indexVectorTuples = new ArrayBuffer[(Int, ColumnVector)]
-            var index = 0
-            isRowDeletedColumn.foreach { columnMetadata =>
-              val isRowDeletedVector = writableVectors(index)
-              rowIndexFilter.get
-                .materializeIntoVector(rowIndex, rowIndex + size, isRowDeletedVector)
-              indexVectorTuples += (columnMetadata.index -> isRowDeletedVector)
-              index += 1
-            }
 
-            rowIndexColumn.foreach { columnMetadata =>
-              val rowIndexVector = writableVectors(index)
-              // populate the row index column value
-              for (i <- 0 until size) {
-                rowIndexVector.putLong(i, rowIndex + i)
+            // When predicate pushdown is enabled we use _metadata.row_index. Therefore,
+            // we only need to construct the isRowDeleted column.
+            if (predicatePushdownEnabled) {
+              isRowDeletedColumn.foreach { columnMetadata =>
+                val isRowDeletedVector = writableVectors.head
+                rowIndexFilter.get.materializeIntoVectorWithRowIndex(
+                  size, batch.column(rowIndexColumn.get.index), isRowDeletedVector)
+                indexVectorTuples += (columnMetadata.index -> isRowDeletedVector)
+              }
+              replaceVectors(batch, indexVectorTuples.toSeq: _*)
+            } else {
+              var index = 0
+              isRowDeletedColumn.foreach { columnMetadata =>
+                val isRowDeletedVector = writableVectors(index)
+                rowIndexFilter.get
+                  .materializeIntoVector(rowIndex, rowIndex + size, isRowDeletedVector)
+                indexVectorTuples += (columnMetadata.index -> isRowDeletedVector)
+                index += 1
               }
 
-              indexVectorTuples += (columnMetadata.index -> rowIndexVector)
-              index += 1
-            }
+              rowIndexColumn.foreach { columnMetadata =>
+                val rowIndexVector = writableVectors(index)
+                // populate the row index column value.
+                for (i <- 0 until size) {
+                  rowIndexVector.putLong(i, rowIndex + i)
+                }
 
-            val newBatch = replaceVectors(batch, indexVectorTuples.toSeq: _*)
-            rowIndex += size
-            newBatch
+                indexVectorTuples += (columnMetadata.index -> rowIndexVector)
+                index += 1
+              }
+
+              val newBatch = replaceVectors(batch, indexVectorTuples.toSeq: _*)
+              rowIndex += size
+              newBatch
+            }
```

Comment:
The existing code should work as is with any `if(predicatePushdownEnabled)`? i.e a new column vector is generated only when `rowIndexColumn` is not empty.